### PR TITLE
[windows] Allow delegation of top-level WindowProc

### PR DESCRIPTION
**Pull request by stuartmorgan**
## Description

Adds APIs for runners to delegate WindowProc handlers into the Flutter
engine, and for plugins to register as possible delegates.

This allows for plugins to alter top-level window behavior in ways that
can only be done from the WindowProc, such as resize control. This
functionality remains entirely on the native side, so is synchronous.

## Related Issues

Part of https://github.com/flutter/flutter/issues/53168

## Tests

I added the following tests: Unit tests for the registration and dispatch objects at both the wrapper and implementation levels.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: shell/platform/windows/client_wrapper/flutter_view_controller.cc [hunk 1]
```diff
@@ -29,6 +29,17 @@ FlutterViewController::~FlutterViewController() {
   }
 }
 
+std::optional<LRESULT> FlutterViewController::HandleTopLevelWindowProc(
+    HWND hwnd,
+    UINT message,
+    WPARAM wparam,
```
Comment (chinmaygarde) [inline shell/platform/windows/client_wrapper/flutter_view_controller.cc:35]: Just a nit but can we give these parameters better names? wparam and lparam is a bit terse. Same elsewhere.

File: shell/platform/windows/client_wrapper/flutter_view_controller.cc [hunk 2]
```diff
@@ -29,6 +29,17 @@ FlutterViewController::~FlutterViewController() {
...
+    LPARAM lparam) {
+  LRESULT result;
+  bool handled = FlutterDesktopViewControllerHandleTopLevelWindowProc(
+      controller_, hwnd, message, wparam, lparam, &result);
+  return handled ? result : std::optional<LRESULT>();
```
Comment (chinmaygarde) [inline shell/platform/windows/client_wrapper/flutter_view_controller.cc:40]: Nit: `std::nullopt` there please. I am never sure whether this means the same same as `std::nullopt` or an optional with a default initialized value. Maybe that's just me.

File: shell/platform/windows/client_wrapper/include/flutter/plugin_registrar_windows.h [hunk 3]
```diff
@@ -35,9 +44,78 @@ class PluginRegistrarWindows : public PluginRegistrar {
...
 
+  // Registers |delegate| to recieve WindowProc callbacks for the top-level
+  // window containing this Flutter instance. Returns an ID that can be used to
+  // unregister the handler.
+  //
+  // Delegates are not guaranteed to be called:
+  // - The application may choose not to delegate WindowProc calls.
+  // - If multiple plugins are registered, the first one that returns a value
+  //   from the delegate message will "win", and others will not be called.
+  //   The order of delegate calls is not defined.
```
Comment (chinmaygarde) [inline shell/platform/windows/client_wrapper/include/flutter/plugin_registrar_windows.h:55]: This seems like an unnecessary restriction that can easily be lifted if the Compare template parameter for the map below was specified. Saying either the first or last one wins ought to be a much less surprising API behavior.

Comment (stuartmorgan) [reply shell/platform/windows/client_wrapper/include/flutter/plugin_registrar_windows.h:55]: It's not that simple. Plugins on Windows are DLLs (to minimize the possibility of problematic interactions), so in practice each plugin has its own instance of this class, and changing the sort behavior here would do nothing (unless a single plugin registered multiple delegates, which is possible, but unlikely). It's actually the sort order in the engine that would matter most of the time. I could define it there, but then there are bizarre edge cases like: plugin A registers a handler, then plugin B registers a handler, then plugin A registers another handler. Without retooling the API to be more complex just to handle that case, the order of calls would actually be A1 A2 B, not A1 B A2.

I also don't really see that there's any value in such a guarantee, because in practice "first" or "last" aren't really under the control of a plugin developer. Assuming people will generally register this on plugin instantiation, that means that if we make this fully defined it devolves to the order of plugin registration. That's not under the control of a plugin developer, and in practice we don't make any guarantees about that. And even if we did (which I actually filed a bug about, because it's annoying if generated code that's checked in changes for no good reason, which happened recently), then we're basically saying that WindowProc handling is decided by alphabetical ordering of plugin names, which is weird, and could have the effect of people picking odd names for plugins just to try to win conflicts.

Assuming anyone actually tested the interaction between their plugin and other plugins, which I think is going to be very rare at best.

**tl;dr**: It's non-trivial added complexity, and I don't think it has any real value.
